crates/pipeline-manager/src/cluster_health.rs

```diff
@@ -0,0 +1,164 @@
+use chrono::{DateTime, Utc};
+use log::{info, warn};
+use reqwest::Client;
+use serde::Serialize;
+use std::{error::Error, sync::Arc, time::Duration};
+use tokio::sync::RwLock;
+use utoipa::ToSchema;
+
+/// Interval between checking health again.
+const POLL_INTERVAL: Duration = Duration::from_secs(10);
+
+/// Default HTTP request timeout to use
+const DEFAULT_REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug, Clone, Serialize, PartialEq, ToSchema, Default)]
+pub struct HealthStatus {
+    pub runner: ServiceStatus,
+    pub compiler: ServiceStatus,
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, ToSchema)]
+pub struct ServiceStatus {
+    pub healthy: bool,
+    pub message: String,
+    pub unchanged_since: DateTime<Utc>,
+}
+
+impl Default for ServiceStatus {
+    fn default() -> Self {
+        Self {
+            healthy: false,
+            message: String::new(),
+            unchanged_since: Utc::now(),
+        }
+    }
+}
+
+async fn poll_service(url: &str, client: &Client, timeout: Duration) -> (bool, String) {
+    match client.get(url).timeout(timeout).send().await {
+        Ok(resp) if resp.status().is_success() => (true, "healthy".to_owned()),
+        Ok(resp) => {
+            let json: serde_json::Value = resp.json().await.unwrap_or_else(
+                |_| serde_json::json!({ "error": "healthz returned invalid JSON" }),
+            );
+            (false, json.to_string())
+        }
+        Err(e) => {
+            let mut error_message = None;
+            if e.is_connect() {
+                let mut err: &dyn Error = &e;
+                while let Some(src) = err.source() {
+                    err = src;
+                }
+                error_message = Some(format!("Failed to connect: {err}"));
+            }
+
+            if e.is_timeout() {
+                let mut err: &dyn Error = &e;
+                while let Some(src) = err.source() {
+                    err = src;
+                }
+                error_message = Some(format!(
+                    "Request timed out (duration {}s): {err}",
+                    timeout.as_secs()
+                ));
+            }
+
+            let error_message = error_message.unwrap_or_else(|| e.to_string());
+            (false, format!("Unreachable: {error_message}"))
+        }
+    }
+}
+
+// Helper function: if the state changed, update timestamp; if not, preserve it.
+fn build_status(ok: bool, msg: String, prev: &ServiceStatus) -> ServiceStatus {
+    let state_changed = prev.healthy != ok && prev.message != msg;
+    ServiceStatus {
```

Comment:
I think unchanged_since is useful, but it is important to know when the last check was. Could you also add `checked_at`?